libraries/AC_AutoTune/AC_AutoTune.cpp

```diff
@@ -1713,56 +1290,831 @@ void AC_AutoTune::get_poshold_attitude(float &roll_cd_out, float &pitch_cd_out,
+    float gyro_reading = 0;
+    switch (test_axis) {
+    case ROLL:
+        gyro_reading = ahrs_view->get_gyro().x;
+        lean_angle = dir_sign * (ahrs_view->roll_sensor - (int32_t)start_angle);
+        break;
+    case PITCH:
+        gyro_reading = ahrs_view->get_gyro().y;
+        lean_angle = dir_sign * (ahrs_view->pitch_sensor - (int32_t)start_angle);
+        break;
+    case YAW:
+        gyro_reading = ahrs_view->get_gyro().z;
+        lean_angle = dir_sign * wrap_180_cd(ahrs_view->yaw_sensor-(int32_t)start_angle);
+        break;
+    }
+
+    // Add filter to measurements
+    float filter_value;
+    switch (tune_type) {
+    case SP_DOWN:
+    case SP_UP:
+        filter_value = dir_sign * (ToDeg(gyro_reading) * 100.0f);
+        break;
+    default:
+        filter_value = dir_sign * (ToDeg(gyro_reading) * 100.0f - start_rate);
+        break;
+    }
+    rotation_rate = rotation_rate_filt.apply(filter_value,
+                    AP::scheduler().get_loop_period_s());
+
+    switch (tune_type) {
+    case RD_UP:
+    case RD_DOWN:
+        twitching_test_rate(rotation_rate, target_rate, test_rate_min, test_rate_max);
+        twitching_measure_acceleration(test_accel_max, rotation_rate, rate_max);
+        twitching_abort_rate(lean_angle, rotation_rate, abort_angle, test_rate_min);
+        break;
+    case RP_UP:
+        twitching_test_rate(rotation_rate, target_rate*(1+0.5f*aggressiveness), test_rate_min, test_rate_max);
+        twitching_measure_acceleration(test_accel_max, rotation_rate, rate_max);
+        twitching_abort_rate(lean_angle, rotation_rate, abort_angle, test_rate_min);
+        break;
+    case SP_DOWN:
+    case SP_UP:
+        twitching_test_angle(lean_angle, rotation_rate, target_angle*(1+0.5f*aggressiveness), test_angle_min, test_angle_max, test_rate_min, test_rate_max);
+        twitching_measure_acceleration(test_accel_max, rotation_rate - dir_sign * start_rate, rate_max);
+        break;
+    default:
+        break;
+    }
+}
+
+void AC_AutoTune::rate_ff_test_init()
+{
+    ff_test_phase = 0;
+    rotation_rate_filt.reset(0);
+    rotation_rate_filt.set_cutoff_frequency(5.0f);
+    command_filt.reset(0);
+    command_filt.set_cutoff_frequency(5.0f);
+    target_rate_filt.reset(0);
+    target_rate_filt.set_cutoff_frequency(5.0f);
+    test_command_filt = 0.0f;
+    test_rate_filt = 0.0f;
+    test_tgt_rate_filt = 0.0f;
+    filt_target_rate = 0.0f;
+    settle_time = 200;
+    phase_out_time = 500;
+}
+
+void AC_AutoTune::rate_ff_test_run(float max_angle_cd, float target_rate_cds, float dir_sign)
+{
+    float gyro_reading = 0.0f;
+    float command_reading = 0.0f;
+    float tgt_rate_reading = 0.0f;
+    const uint32_t now = AP_HAL::millis();
+    static float trim_command_reading = 0.0f;
+    static float trim_heading = 0.0f;
+    static float rate_request_cds;
+    static float angle_request_cd;
+
+    // TODO make filter leak dependent on dt
```

Comment:
Does this need to be done before this is merged?

Reply:
done